### ISSUE TITLE
Parser error reading Open@Easy : null

### ISSUE THREAD

**Issue by mok33** [jpickup/GarminTools#6]
Hi again! 
I want to add an open step to my workout. It doesn't take it into account (parsing error). I am using the Windows installer.
Here is my xls file that I am trying to convert to FIT.
[workout.xls](https://github.com/jpickup/GarminTools/files/12790742/workout.xls)

**Comment by jpickup** [comment]
I'll take a look and get back to you with either a revised XLS or updated code.

**Comment by jpickup** [comment]
Sorry, I'm a bit confused. Could you elaborate where you see an error and what it is? I've tried loading the workout.xls above and I see FIT files successfuly created. That said, I don't see any open steps in the sheet. By an open step I assume you mean one that only ends when the lap button is pushed, in other words no time or distance target. Some examples of open workouts:

Open
Open@Z3
Open@140-170bpm
Open@Slow

**Comment by jpickup** [comment]
As a further test I added a row to workout.xls with "Open@Easy" and then tried to load, here's what I see:
![image](https://github.com/jpickup/GarminTools/assets/1729293/15d28350-3cd2-43a7-94ca-64135f435d64)

**Comment by mok33** [comment]
Hi, I apologize for the bad xls I provided. I tried with your xls file I get the same parsing issue with Open: 
`Converting C:\Users\emohamed\workout\TestWorkouts.xls
Converting C:\Users\emohamed\workout\TestWorkouts.xls to FIT files in C:\Users\emohamed\workout
Writing output to C:\Users\emohamed\workout
Reading workout schedule
Workbook (readfile) created with reclen=497
found SupBook record at 301
exit create workbook from existing file function
convertLabelRecords called
convertLabelRecords exit
sheet createSheet (existing file) exited
Time at start of cell creating in HSSF sheet = 1696329089183
total sheet cell creation took 7ms
sheet createSheet (existing file) exited
Time at start of cell creating in HSSF sheet = 1696329089191
total sheet cell creation took 0ms
Parser error reading Open : null
Error converting C:\Users\emohamed\workout\TestWorkouts.xls
Error parsing Open
`

**Comment by jpickup** [comment]
OK, this is interesting! Same file doesn't work on a different machine. There are a few possibilities here:

1. As you mention, Windows installer not up-to-date
2. Some difference in Mac (where I test) vs Windows
3. A localisation difference, e.g. something only works for my GB locale

The first two I can check fairly easily and will do so tonight. #3 will be tricky so I really hope it's not that.

**Comment by jpickup** [comment]
It's the first one. Open steps were only added on 26-Jan-2023 and the Windows installer package predates that. I will build a new windows install package tonight and upload it. Sorry about this!

**Comment by jpickup** [comment]
The new Windows installer msi is now checked in to GitHub.  For some reason I needed to uninstall the app in order to be able to re-run the installer. Not sure why but I'm not really clued up on Windows.
Let me know if you have any more problems.

**Comment by mok33** [comment]
thank you I tried the open command its working now ! 
I have another issue however when I try to repeat several steps multiple times using (a + b +c + d) * n, only the first two step are repeated. so only (a + b) * n.
Here is the xls to reproduce the issue :
[workoutWRIN.xls](https://github.com/jpickup/GarminTools/files/12801873/workoutWRIN.xls)

**Comment by jpickup** [comment]
I'll take a look. 
Thanks.

**Comment by jpickup** [comment]
I've reporoduced the problem and understand it, which is good. Now just need to figure out a solution as the problem is in how the Antlr grammar generates Java during the parsing. It's one of those things that'll be a one line change that's easy once you've figured it out.

**Comment by jpickup** [comment]
I'm pretty sure I've fixed that bug. Please try the latest Windows installer. 
(it's bad that fixing this was more interesting than my real job!)

**Comment by mok33** [comment]
it's working flawlessly now 🤩!! 
You're the best!

**Comment by jpickup** [comment]
That's great news. I'm really happy that this is useful to you as you are most likely the only other person using this. Let me know if there are any features you would like adding. I realise that the user interface isn't pretty, that's one area where my skills aren't great.

**Comment by mok33** [comment]
I shared your app with my running group (100+ runners)! You might have new users ^^

**Comment by Abasz** [comment]
Though I am not using this directly but it served as a very good basis to the app I developed :) So thank you for your effort :)

**Comment by mok33** [comment]
Hi John, 
so yesterday when tried the workout (see last csv), for some reason, my (3'-2'-1') * 2 became a 3'-2'-1'-1'-2'-3'. The second set was in reverse (1-2-3 instead of 3-2-1).

**Comment by jpickup** [comment]
I'll have a look. This could be an interesting thing to understand/diagnose. The strange this is that what is sent to the watch in the FIT file is a collection of steps and then a repeat count, so "(A + B + C) * n" really does get sent like that it's not converted to A + B + C + A + B + C

**Comment by jpickup** [comment]
I'm struggling to reproduce this. Please could you double check? Here's what I did, taking this workout:
`Open@Easy + (3:00@VO2Max90 +   1:00@VeryEasy + 2:00@VO2Max95 + 1:00@VeryEasy + 1:00@VO2Max + 1:00@VeryEasy)   * 2 + Open@VeryEasy`
and it loading onto the watch if I then start a new activity and select the workout and view steps, scrolling though I see 9 steps as follows:
1. Run Until Lap Press 9:49 - 10:44/mi  (sorry my watch is in Miles not km)
2. Run 3:00 6:50-7:02/mi
3. Run 1:00 10:12-13:25/mi
4. Run 2:00 6:42-6:50/mi
5. Run 1:00 10:12-13:25/mi
6. Run 1:00 6:31-6:39/mi
7. Run 1:00 10:12-13:25/mi
8. Repeat 2-7 2x
9. Run Until Lap Press 10:12-13:25/mi

This seems to line up with what's in the workout text.

Can you try doing this and seeing what Garmin shows on your watch?

What I've not tried to do is actually go out an run the workout but if what Garmin displays here versus what happens on the run is different then that will be evern more interesting to track down!

**Comment by mok33** [comment]
So my Garmin shows me the same thing as you. I just checked yesterday's laps (see image below). laps 1-4 are step 1 (Run Until Lap) Everything is okay until step 11 I think it was my fault I think I misclicked the lap button instead of pausing (I was out of breath hahaha)
So I am going to close the issue! I apologize.
![image](https://github.com/jpickup/GarminTools/assets/31034940/89fe5684-45c7-4da4-8706-ead8c0e58cca)
